src/indexsets.jl

```diff
@@ -0,0 +1,78 @@
+abstract type AbstractSparsityPattern end
+#=
+Expected interface:
+  * myempty(::Type{MyPattern}): return an empty pattern
+  * seed(::Type{MyPattern}, i::Integer): return an pattern that only contains the given index `i`
+  * Base.show
+
+And for compatibility with...
+  * ConnectivityTracer: inputs(p::MyPattern)   
+  * GradientTracer:     gradient(p::MyPattern) 
+  * HessianTracer:      gradient(p::MyPattern) AND hessian(p::MyPattern) 
+=#
+
+# Utilities on AbstractSet
+myempty(::Type{S}) where {S<:AbstractSet} = S()
+seed(::Type{S}, i::Integer) where {S<:AbstractSet} = S([i])
+
+## First order
+abstract type AbstractFirstOrderPattern <: AbstractSparsityPattern end
+
+"""
+$(TYPEDEF)
+
+First order sparsity pattern represented by an index set of non-zero values.
+Represented by an `AbstractSet` of integer indices.
+
+## Fields
+$(TYPEDFIELDS)
+"""
+struct SetIndexset{S<:AbstractSet{<:Integer}} <: AbstractFirstOrderPattern
```

Comment:
Since I anticipate this weird struct name will get a comment:

This is a mathematical index set represented by a Julia `AbstractSet` ($\rightarrow$ `SetIndexset`).
As mentioned in the PR description, I plan on adding an index set represented by a Julia `Dict` ($\rightarrow$ `DictIndexset`) in a follow-up PR.

Reply:
In my mind, a mathematical index set is just one specific way to represent a sparsity pattern.

Reply:
Sure but just name it something that ends with `Pattern`, otherwise this gets very confusing when you bring it together with _actual_ set types. `IndexSetFirstOrderPattern` is fine

Reply:
In my eyes, renaming an `Indexset` to `Pattern` is just as non-informative as renaming `Float32` to `Number`.

How about `IndexSet` and `SecondOrderIndexSet`?

Reply:
Keep in mind that you will want to add some sort of `SharedIndexSet` in (#107). This is why I wanted to differentiate this from `IndexSet` by using `SetIndexset`.

Reply:
> How about IndexSet and SecondOrderIndexSet?
Or `SimpleIndexSet`, since using `AbstractSet`s is the most natural way to implement them? But now we don't even restrict to `AbstractSet` anymore...

Reply:
https://en.wikipedia.org/wiki/Index_set

Reply:
The pattern can be represented by an https://en.wikipedia.org/wiki/Index_set, which is what we currently do. An alternative would be a mask like a BitMatrix.
Both are very different kinds of sparsity pattern representations.

Reply:
I know!
The first kind should be called IndexSetPattern, the second kind BitMatrixPattern.

Reply:
If you didn't want such long names, you shouldn't have introduced a third layer between tracers and sets

Reply:
Ok, maybe I misunderstood, I thought you were arguing for shorter names.

Reply:
> The first kind should be called IndexSetPattern, the second kind BitMatrixPattern.

The issue with this is that *mathematical index sets* can be implemented using actual (a) sets, (b) dicts, and (c) shared datastructures. We will therefore end up with 

* a) `SetIndexSetPattern`
* b) `DictIndexSetPattern`
* c) `SharedIndexSetPattern`

Reply:
So my suggestion is so call these
* a) `SimpleIndexSet`
* b) `DictIndexSet`
* c) `SharedIndexSet`

The way we don't call `Float32`s `Float32Number`s.

Reply:
My issue is that when people see something like `SimpleIndexSet`, they might expect it to behave like an `AbstractSet`. But here it's something very different, subtyping `AbstractPattern` and _containing_ one or two set objects.
To correct your analogy, it's as if we defined a new type `BitNumber <: AbstractVector{Bool}`. Sure, a number is a vector of bits, but when people see `BitNumber`, they will expect it to behave like a `Number`, not like an `AbstractVector`.
In the same way, when people see `IndexSet`, they will expect it to behave like a `Set`, not like an `AbstractPattern`. Hence the need to add `Pattern` at the end, and it doesn't hurt anyone.

Reply:
Now I see your point. 👍